[WIP] Add mypy and static type checking to Habitat Lab

Skylion007 opened the pull request:
## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
Actually uses all the TypeHints we have in Habitat-Lab by adding a pre-commit hook for them.
Before doing that, there are quite a lot of type signature errors that need to be fixed in habitat. 

This actually required quite a lot of refactoring just to get the base habitat to use properly use these type hints and to fix existing errors in typing. I had to add quite a few "ignores" to fix some pretty serious issues we have with inheritance and overloading variables / types of Python classes.

As of this typing, I have fixed all except the eqa_task and tasks/nav/nav.py. These two classes alone account for 84 typing errors in Habitat, currently.


<!--- Please link to an existing issue here if one exists. -->
<!--- (we recommend to have an existing issue for each pull request) -->

## How Has This Been Tested
With PyTest and MyPy
Like with Habitat-Sim I used monkeytype to help out, but since so many existing type hints were wrong and the fact that we have class names that are the same throughout habitat, it required quite a large number of manual changes so far.


<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!--- What types of changes does your code introduce? Leave all the items that apply: -->
- Large Refactor and Dev Tools update
## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [X] All new and existing tests passed.

Skylion007 commented:
@erikwijmans @mathfac a huge pain problem here is the our core classes do not defined any call signatures which is making overloading them quite annoying as I have to add ignores all for every single sensor or measure. We should standardized a Sensor / Measurement API to fix this.  See incompatible overrides for more info: https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides

Skylion007 commented:
I added enough updated typehints and ignore comments for the bare minimum of it to stop complaining. We tend to extend classes in very non-pythonic ways around Habitat-Lab.

We also should really fix the call signature of the base Sensor and Measure class so that it's more consistent.

erikwijmans commented on habitat/core/registry.py:
```diff
@@ -193,35 +206,37 @@ def register_action_space_configuration(
 
     @classmethod
-    def get_simulator(cls, name):
+    def get_simulator(cls, name: str) -> Type["HabitatSim"]:
```
IIRC, the PyRobot simulator is also in the registry, so this should just be `Type["Simulator"]`

Skylion007 replied:
Good catch.

erikwijmans reviewed (changes requested):
I get the motivation of doing this, but right now there are so many `# type: ignore`'s floating around that I am not sure if the value add vs. PITA ratio is right.

Maybe we start with typing and enforcing typing in at least `habitat/core` and maybe `habitat` and slowly expand it?

Skylion007 commented on habitat/tasks/nav/object_nav_task.py:
```diff
@@ -125,39 +132,42 @@ def _get_observation_space(self, *args: Any, **kwargs: Any):
-            max_value = max(
-                self._dataset.category_to_task_category_id.values()
+            max_value = (
+                max(self._dataset.category_to_task_category_id.values()),
```
@erikwijmans It is a tuple, mypy caught this inconsistency. (see the assignment two lines above it).

erikwijmans replied:
Ah, the assignment two lines above is wrong then.  It doesn't make sense that high would a tuple when low is an int

Skylion007 commented on habitat/datasets/object_nav/object_nav_dataset.py:
```diff
@@ -62,23 +62,24 @@ def to_json(self) -> str:
-            self.episodes[i].goals = self.goals_by_category[
-                self.episodes[i].goals_key
-            ]
+            self.episodes[i].goals = list(
```
Because goals can be a sequence with the new DatasetClass so it is needed to narrow the type.

erikwijmans replied:
That doesn't seem like a good reason to add needless copies tho.  Why can't goals be a list in the dataset class?

Skylion007 replied:
Because it's covariant currently.

Skylion007 commented:
@mosra Really confused how this broke our doc build. Any ideas?

mosra commented:
Heh. Looks like the newly added hints hit some corner case in the doc generator, and it might be some Py3.6-specific thing, even. I suppose you don't have the doc build running locally so you could run `python.py` with `--debug` to see at least what class/module it blows up on?

I'll check this tomorrow, please ping me on Slack in case I forget ;)

mosra commented:
Found the culprit and made a workaround in m.css, it's something deep inside undocumented Python `typing` internals again. Currently fighting some CI breakage annoyances, hopefully I'll have it landed in m.css `master` later today.

mosra commented:
@Skylion007 Fix is in mosra/m.css@0b94575561f7f1ecc19642e7f3ca680cc71d6eac. Pull the `doc/m.css` submodule in `habitat-sim` to get it.

Skylion007 commented:
@erikwijmans Any further reduction in type ignores requires us to remove most of the args / kwargs passing in `habitat.core`

Skylion007 commented:
@erikwijmans Great news, a version of MyPy just dropped yesterday that allows me to do global/module level ignores just like Flake8 does. This should solve the issue. I also went back and removed most of the type ignores and replaced with explicit PEP528 casting where I could or defined the missing attributes on the parent objects.

erikwijmans commented on habitat_baselines/rl/ppo/ppo.py:
```diff
@@ -4,28 +4,37 @@
-        use_clipped_value_loss=True,
-        use_normalized_advantage=True,
-    ):
+        actor_critic: Union[PointNavResNetPolicy, PointNavBaselinePolicy],
```
```suggestion
        actor_critic: Policy,
```

The base class should work here

Skylion007 commented:
Whoops, didn't mean to push that.

mathfac reviewed (approved):
Thank you Aaron for improving checks on  API promises. Looks great to go.

Skylion007 commented:
Grr, just realized a bug where all the Habitat_Sim type imports now block a soft install of habitat_lab without habitat_sim. We need to create a CI Test for having habitat-lab work without habitat_sim. Throughts @erikwijmans @mathfac ?

Skylion007 commented:
Fixed in #498
